### PR TITLE
Moved Frame-Pointer Validation to GRPC

### PR DESCRIPTION
**Pull request by florian-kuebler**
Moved frame-pointer-validator service and UI from Asio to gRPC.
Also reduces state to be present for the validation on the server (module information/functions) and instead send the functions to be loaded directly (offset and size). 

Dissassembling requires the bit'ness of the process (i.e. 64 or 32 bit). This information is available on the client, but we do not want to rely on this, as the service should be able to be used e.g. as a static analysis at some point later. The bit'ness is part of the ElfHeader and as we have the modules path, we can easily read it. This PR extends the ElfFile with a method to retrieve that info.

### CODE REVIEW

File: OrbitService/FramePointerValidatorServiceImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+#include "FramePointerValidator.h"
+#include "OrbitFunction.h"
+
+grpc::Status FramePointerValidatorServiceImpl::ValidateFramePointers(
+    grpc::ServerContext*, const ValidateFramePointersRequest* request,
+    ValidateFramePointersResponse* response) {
+  for (const auto& module_info : request->modules()) {
+    // Even though this information should be available on the client,
+    // we want not rely on this here, and for this particular use case we are
+    // fine with doing some extra work, and compute it here.
```
Comment (dimitry-) [inline OrbitService/FramePointerValidatorServiceImpl.cpp:19]: read it from the elf-file

File: protos/frame-pointer-validator.proto [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+message ModuleInformation {
+  string module_path = 1;
+  repeated FunctionInformation functions = 2;
+}
+
+message FunctionInformation {
+  uint64 offset = 1;
+  uint64 size = 2;
+}
```
Comment (dimitry-) [inline protos/frame-pointer-validator.proto:15]: you have this and also there is FunctionInfo struct in the code. Is it necessary to have both? can't you use just this one?

File: protos/frame-pointer-validator.proto [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+// found in the LICENSE file.
+
+syntax = "proto3";
+
+message ModuleInformation {
+  string module_path = 1;
+  repeated FunctionInformation functions = 2;
+}
+
+message FunctionInformation {
```
Comment (dimitry-) [inline protos/frame-pointer-validator.proto:12]: Maybe CodeBlock?

File: protos/frame-pointer-validator.proto [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+syntax = "proto3";
+
+message ModuleInformation {
+  string module_path = 1;
+  repeated FunctionInformation functions = 2;
```
Comment (dimitry-) [inline protos/frame-pointer-validator.proto:9]: move these to request.

File: protos/frame-pointer-validator.proto (hunk 2 above)
Comment (florian-kuebler) [reply protos/frame-pointer-validator.proto:15]: I wanted to avoid using the FunctionInfo struct, which gets generated by the proto.

Comment (ronaldfw) [reply protos/frame-pointer-validator.proto:15]: Unless we have good reasons for doing so, we shouldn't duplicate protos as structs. (A good reason would be a measurable performance difference.)

Comment (florian-kuebler) [reply protos/frame-pointer-validator.proto:15]: The `FunctionInfo` struct is used in `OrbitCore/FramePointerValidator` and OrbitCore is not linked against OrbitProtos. Thus, I would either need to link them, or stick with the struct.

File: protos/frame-pointer-validator.proto (hunk 3 above)
Comment (florian-kuebler) [reply protos/frame-pointer-validator.proto:12]: You mean as a name?

File: protos/frame-pointer-validator.proto (hunk 4 above)
Comment (florian-kuebler) [reply protos/frame-pointer-validator.proto:9]: But how would I do this? Both contain repeated fields.

File: protos/frame-pointer-validator.proto (hunk 3 above)
Comment (dimitry-) [reply protos/frame-pointer-validator.proto:12]: yep.

File: protos/frame-pointer-validator.proto (hunk 4 above)
Comment (dimitry-) [reply protos/frame-pointer-validator.proto:9]: Let's do one request per module - let the client generate multiple requests if multiple modules need to be validated. This also makes error handling easier.

File: protos/frame-pointer-validator.proto (hunk 2 above)
Comment (dimitry-) [reply protos/frame-pointer-validator.proto:15]: You can link it against proto.

File: protos/frame-pointer-validator.proto (hunk 4 above)
Comment (florian-kuebler) [reply protos/frame-pointer-validator.proto:9]: The issue with this, is that I would then need to keep track of those in the UI and aggregate the results there.

File: protos/frame-pointer-validator.proto (hunk 3 above)
Comment (florian-kuebler) [reply protos/frame-pointer-validator.proto:12]: okay. thanks.

File: protos/frame-pointer-validator.proto (hunk 4 above)
Comment (florian-kuebler) [reply protos/frame-pointer-validator.proto]: Nevermind, I changed it that way.

File: protos/frame-pointer-validator.proto (hunk 2 above)
Comment (florian-kuebler) [reply protos/frame-pointer-validator.proto:10]: done. thanks.

File: OrbitGl/FramePointerValidatorClient.cpp [hunk 5]
```diff
@@ -75,6 +68,6 @@ void FramePointerValidatorClient::HandleResponse(const Message& message,
 
   std::string text =
       absl::StrFormat("Failed to validate %d out of %d functions",
-                      std::get<1>(response).size(), num_functions);
+                      num_fpo_functions, num_functions);
```
Comment (dimitry-) [inline OrbitGl/FramePointerValidatorClient.cpp:71]: what if all functions are compliant? consider showing something more positive in this case. Failed to validate 0 out of 10000 functions does not sound as a success.

Maybe change the message to something like
"Validation complete
Module \"%s\": %d functions support frame pointers, %d functions don't.
...."

File: protos/frame-pointer-validator.proto [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
Comment (dimitry-) [inline protos/frame-pointer-validator.proto:1]: rename it to code_block.proto?

File: OrbitService/FramePointerValidatorServiceImpl.cpp [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  for (const auto& function : functions.value()) {
+    CodeBlock* added_function = response->add_functions_without_frame_pointer();
+    added_function->set_offset(function.offset());
+    added_function->set_size(function.size());
+  }
+
+  return grpc::Status::OK;
+}
+
+//
```
Comment (dimitry-) [inline OrbitService/FramePointerValidatorServiceImpl.cpp:43]: ?

File: OrbitGl/FramePointerValidatorClient.cpp (hunk 5 above)
Comment (florian-kuebler) [reply OrbitGl/FramePointerValidatorClient.cpp]: Renamed it, however I kept the summary instead of a dialogue for each module.

**Comment by florian-kuebler** [comment]
@dimitry- Addressed the review comments. Could you approve once more, as soon as you looked at the new message we are showing?

File: OrbitGl/FramePointerValidatorClient.cpp [hunk 8]
```diff
@@ -11,70 +11,64 @@
...
     num_functions += module->m_Pdb->GetFunctions().size();
   }
 
-  std::string text =
-      absl::StrFormat("Failed to validate %d out of %d functions",
-                      std::get<1>(response).size(), num_functions);
+  std::string text = absl::StrFormat(
+      "Validation complete.\n%d functions support frame pointers, %d functions "
+      "don't.",
+      num_functions - num_fpo_functions, num_fpo_functions);
```
Comment (dimitry-) [inline OrbitGl/FramePointerValidatorClient.cpp:72]: I think it is more useful to have per-module numbers. wdyt?

File: OrbitGl/FramePointerValidatorClient.cpp (hunk 5 above)
Comment (florian-kuebler) [reply OrbitGl/FramePointerValidatorClient.cpp]: After discussed offline, I renamed it again, to give the overview per module.

File: OrbitGl/FramePointerValidatorClient.cpp [hunk 9]
```diff
@@ -58,17 +59,14 @@ void FramePointerValidatorClient::AnalyzeModules(
...
-  for (const auto& module : modules) {
-    num_functions += module->m_Pdb->GetFunctions().size();
-  }
-
-  std::string text = absl::StrFormat(
-      "Validation complete.\n%d functions support frame pointers, %d functions "
-      "don't.",
-      num_functions - num_fpo_functions, num_fpo_functions);
+  std::string text =
+      absl::StrJoin(dialogue_messages.begin(), dialogue_messages.end(), "\n");
```
Comment (dimitry-) [inline OrbitGl/FramePointerValidatorClient.cpp:70]: I think you can just pass dialogue_messages here, no need to use iterators.

Comment (florian-kuebler) [reply OrbitGl/FramePointerValidatorClient.cpp:70]: Cool, thanks

**Pull request merged**